Fix timeline semaphore value difference validations on binary semaphores

farnoy opened the pull request:
Fixes #1786 

The scenario that was breaking was where you had binary semaphores in one bucket (wait/signal), but you had at least one timeline semaphore in the other. Having all semaphores be binary skips the `*SemaphoreValueCount` validation here: https://github.com/KhronosGroup/Vulkan-ValidationLayers/blob/2f609275c00c66f4258f9bd07217d47d58a2f73b/layers/core_validation.cpp#L2127-L2134

And rightfully so, because the spec states:
> [...] and **any** element of pSignalSemaphores was created with a VkSemaphoreType of VK_SEMAPHORE_TYPE_TIMELINE, then its signalSemaphoreValueCount member must equal signalSemaphoreCount

But, when that is skipped, the validation code does not bail out early and is allowed to fall through to value difference checks, which assume that the value count parameter was validated.

To address the false positive message about value difference on binary semaphores, I added a guard inside of `ValidateMaxTimelineSemaphoreValueDifference`. We already use semaphore state there, and `PreCallValidateSignalSemaphore` (one of the call sites) verifies semaphore type on its own, so it should be a safe addition.

To prevent SEGFAULTs on a null pointer, I've added guards outside of function calls to check for ValueCount. I considered checking directly for the null pointer, but I think parameter validation does it already and the call sites operate on count arguments.

An alternative implementation of `ValidateMaxTimelineSemaphoreValueDifference` could accept a lambda to get the payload, instead of always being provided a payload. That kind of laziness would let us defer using that value until it's necessary, avoiding a segfault and a guard outside the function call.

The test suite here fails for me with `SIGFPE` in `[ RUN      ] VkLayerTest.CreatePipelineExceedMaxTessellationControlInputOutputComponents`, so I wasn't able to add a regression test. I'll do a second attempt later.

jekstrand commented on layers/core_validation.cpp:
```diff
@@ -2338,16 +2341,20 @@ bool CoreChecks::PreCallValidateQueueSubmit(VkQueue queue, uint32_t submitCount,
-            }
+            // If there are any timeline semaphores, this condition gets checked before the early return above
+            if (info->waitSemaphoreValueCount)
+                for (uint32_t i = 0; i < submit->waitSemaphoreCount; ++i) {
```
This doesn't do anything useful.  If `info->waitSemaphoreValueCount` is zero, then the  loop will execute zero times.

farnoy replied:
These are separate variables and you can call `vkQueueSubmit` with `info->waitSemaphoreValueCount == 0` and `submit->waitSemaphoreCount > 0`. As long as all wait semaphores are binary, the VUID you quoted is not violated and this PreCall validation hook does not bail out early.

jekstrand replied:
Sorry.  Not fully awake yet.  You're right that they are two different variables.  Would it be better to just change the loop bounds to use `waitSemaphoreValueCount`?

farnoy replied:
Yes, that should be fine. If there are any timeline semaphores in there, `waitSemaphoreValueCount == waitSemaphoreCount` will be asserted by the earlier code. I'll make this change

jekstrand commented:
It'd also be good to add a validation check for the VU that I quoted in #1786:

> If the pNext chain of this structure includes a VkTimelineSemaphoreSubmitInfo structure and any element of pWaitSemaphores was created with a VkSemaphoreType of VK_SEMAPHORE_TYPE_TIMELINE, then its waitSemaphoreValueCount member must equal waitSemaphoreCount

That would have caught the bug in the blog post example.

farnoy commented:
@jekstrand The snippet I referenced in the description is validating precisely the VUID you're quoting: `VUID-VkSubmitInfo-pNext-03240` https://www.khronos.org/registry/vulkan/specs/1.2-extensions/html/vkspec.html#VUID-VkSubmitInfo-pNext-03240

The key part that makes this tricky is "and **any** element of pWaitSemaphores was created with a VkSemaphoreType of VK_SEMAPHORE_TYPE_TIMELINE"

jekstrand commented:
You're right.  Again, not enough caffine this morning. :-)

farnoy commented:
I applied a patch to simplify it according to Jason's remark. It is possible to segfault still, but in a different case.

I traced my test app and it was passing `waitSemaphoreValueCount = 1` with `waitSemaphoreCount = 0`. Obviously, this is nonsensical usage (allowed by the spec though), but I was only doing this to appease an earlier version of validation layers and the SDK.

This begs the question: how resilient do we want to be? A scenario that was previously adjusted to pass validation layers now segfaults.

I'm leaning towards deleting that last commit in favor of the older if-guard. It's longer, but does not segfault when called in this weird, but compliant way.

mark-lunarg commented:
@farnoy, sorry about sitting on this one so long -- I thought we were waiting on something.  If you can finalize things and add a test case or two, we'll get things moving forward.

zhaijialong commented:
Any update on this ? I think we met the same issue.

farnoy commented:
I'll try to update it this week with a regression test case

tbiedert commented:
We're also hitting this. Thanks for looking into it!

mark-lunarg commented:
Any updates here?  @Tony-LunarG, @farnoy, do y'all need anything?

farnoy commented:
Hey folks, sorry about the delay here, it was a busy period for me. I finally found some free time and updated this PR. It's now rebased and I've added a regression testing snippet to the existing test `VkLayerTest.InvalidSignalSemaphoreValue`.

This new snippet fails with the following error when ran against the master branch:

```sh
# from my branch, in cmake build dir:
$ git restore --source=origin/master ../layers/core_validation.cpp
$ ninja vk_layer_validation_tests && ./tests/vk_layer_validation_tests --gtest_filter="VkLayerTest.InvalidSignalSemaphoreValue"
```

Outputs:

> Expected to succeed but got error: Validation Error: [ VUID-VkSubmitInfo-pSignalSemaphores-03244 ] Object 0: handle = 0x70000000007, type = VK_OBJECT_TYPE_SEMAPHORE; | MessageID = 0x7fd092df | VkQueueSubmit: value exceeds limit regarding current semaphore VkSemaphore 0x70000000007[] payload The Vulkan spec states: For each element of pSignalSemaphores created with a VkSemaphoreType of VK_SEMAPHORE_TYPE_TIMELINE the corresponding element of VkTimelineSemaphoreSubmitInfo::pSignalSemaphoreValues must have a value which does not differ from the current value of the semaphore or the value of any outstanding semaphore wait or signal operation on that semaphore by more than maxTimelineSemaphoreValueDifference (https://www.khronos.org/registry/vulkan/specs/1.2-extensions/html/vkspec.html#VUID-VkSubmitInfo-pSignalSemaphores-03244)

mark-lunarg commented:
@Tony-LunarG, after CI is happy, please remember to squash the first two commits and fix the commit messages before pushing.  Thanks!

mark-lunarg commented:
@farnoy, can you share any thoughts related to issue #1999?

Tony-LunarG commented:
CI fails with crash in QueueSubmitTimelineSemaphoreBadValue

core_validation.cpp
2664        for (uint32_t i = 0; i < info->waitSemaphoreValueCount; ++i) {
2665                VkSemaphore semaphore = submit->pWaitSemaphores[i];

submit->pWaitSemaphores is 0
info->waitSemaphoreValueCount is 1
submit->waitSemaphoreCount is 0

@farnoy can you please take a look?

farnoy commented:
@Tony-LunarG Can you revert `d48a696122c16968dc16aaa4d256d6f45c613c31` and push to my fork? I believe this fails due to simplifications I made in response to comments from Jason, the code is less resilient right now. I'm on mobile now and going to bed. If not I'll update this tomorrow.

farnoy commented:
Is there a separate CI besides the one that shows up here on GitHub?

mark-lunarg commented:
@farnoy, yeah, Github CI uses no hardware, just builds and tests against a null driver on Linux.  Our internal test farm uses a wide array of real hardware, and we try to make sure everything that ends up in the tree has passed both.
